### ISSUE TITLE
Keras-MXNet custom loss with slice operation loses shape information

### ISSUE THREAD

**Issue by sandeep-krishnamurthy** [awslabs/keras-apache-mxnet#120]
If you are using a Custom Loss function and if you use a slice operation in custom loss function, you need to explicitly set the shape of the loss tensor before returning from your custom loss function.

Example:

```
# Your custom loss function.
def my_custom_loss(y, y_hat):
    # Assume below shapes for demonstrating the issue
    # y.shape -> (None, 10)
    # y_hat.shape -> (None, 10) 
    
    y = y[:, :8]
    y_hat = y_hat[:, :8]
   
   # You used a slice operation. Shape information of y and y_hat is lost after this operation.
   # y.shape -> ()
   # y_hat.shape -> ()

    loss = y - y_hat
    
    # loss.shape -> ()  => This causes custom loss calculation to fail with shape mismatch errors.

    ##### This is a workaround required for MXNet backend due to issue in MXNet - https://github.com/apache/incubator-mxnet/issues/11349 #####
   # Explicitly set the shape of loss tensor before returning. Since first dimension is batch_size you can leave it as 'None' i.e., unknown.
    loss._keras_shape = (None, 8)  
    return (loss)
```

# Reason

As described in this issue https://github.com/apache/incubator-mxnet/issues/11349 MXNet has a strange behavior for slice operator where infer_partial_shape() call returns pre-maturely even if one of the dimension of input is None (unknown). Since y, y_hat batch size is unknown in our case in Keras, after slicing, shape info is lost.

# Solution

Working with MXNet team to resolve the issue and allow infer_shape_partial() to try to come up with shape information.

**Comment by sandeep-krishnamurthy** [comment]
Fix is merged onto MXNet. Resolving.
